core/src/main/scala/org/apache/spark/sql/delta/optimizer/OptimizeMetadataOnlyDeltaQuery.scala

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.optimizer
+
+import org.apache.spark.sql.{Column, Row, SparkSession, functions}
+import org.apache.spark.sql.catalyst.expressions.{Alias, AttributeReference, ExprId, Literal, SubqueryExpression}
+import org.apache.spark.sql.catalyst.expressions.aggregate.{AggregateExpression, Count}
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.delta.DeltaTable
+import org.apache.spark.sql.delta.files.TahoeLogFileIndex
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+import org.apache.spark.sql.delta.stats.DeltaScanGenerator
+import org.apache.spark.sql.functions.count
+import org.apache.spark.sql.types.LongType
+
+trait OptimizeMetadataOnlyDeltaQuery {
+  def optimizeQueryWithMetadata(spark: SparkSession, plan: LogicalPlan): LogicalPlan = {
+    if (!spark.sessionState.conf.getConf(DeltaSQLConf.DELTA_OPTIMIZE_METADATA_QUERY)) {
+      plan
+    } else {
+
+      def transformSubqueries(plan: LogicalPlan): LogicalPlan = {
+        import org.apache.spark.sql.delta.implicits._
+
+        plan transformAllExpressionsUp {
+          case subquery: SubqueryExpression =>
+            subquery.withNewPlan(transform(subquery.plan))
+        }
+      }
+
+      def transform(plan: LogicalPlan): LogicalPlan =
+        transformSubqueries(plan) transform {
+          case CountStarDeltaTable(aliasName, exprId, qualifier, countValue) =>
+            createLocalRelationPlan(aliasName, exprId, qualifier, countValue)
+        }
+
+      transform(plan)
+    }
+  }
+
+  protected def getDeltaScanGenerator(index: TahoeLogFileIndex): DeltaScanGenerator
+
+  private def createLocalRelationPlan(
+    aliasName: String,
+    exprId: ExprId,
+    qualifier: Seq[String],
+    rowCount: Long): LogicalPlan = {
+    val relation = LocalRelation.fromExternalRows(
+      output = Seq(AttributeReference(aliasName, LongType)(exprId, qualifier)),
+      data = Seq(Row(rowCount)))
+
+    relation
+  }
+
+  object CountStarDeltaTable {
+
+    def unapply(plan: Aggregate): Option[(String, ExprId, Seq[String], Long)] = {
+      plan match {
+        case Aggregate(
+        Nil,
+        Seq(oldAlias@Alias
+          (AggregateExpression(Count(Seq(Literal(1, _))), _, false, None, _), aliasName)),

```

Comment:
Will `AggregateExpression.mode` always be the same value? If so, can we use the fixed value instead of `_`

Reply:
I'm not familiar with all the modes. In my tests it is always Complete